Step3 로또(자동) 리뷰 요청드립니다!

ilmare-cbk opened the pull request:
안녕하세요! 우테캠 Pro 5기 교육생 천보경입니다😃
이번 미션은 지난 미션들보다 요구사항이 복잡해서 TDD로 작업하는 것이 꽤 어려웠던 미션이었습니다!
가능한한 작은 요구사항들로 나눠 진행하고자 노력했고,
최대한 모델에 메세지를 보내서 모델 내에서 로직을 처리하도록 구현했습니다! (확실히 이렇게 코딩하다보니 getter 사용이 거의 없었습니다!)
또 메세지를 처리하는 담당이 어떤 모델이어야 하나 많은 고민을 했던 것 같습니다! 😃😃

이번 미션하면서 한가지 여쭤보고 싶은 사항이 있어서 질문 남깁니다!
질문) TDD로 작성하다보니 모델의 메소드 접근제어자가 public인 상태입니다. 하지만 실제로 구현을 다하고 보니, 해당 메소드는 본인 모델 내에서만 사용하고 있었습니다! (테스트에서 사용하는 것 제외) 결국 테스트를 위해 public으로 된 모습과 동일한데, 테스트를 위해  public으로 공개하는것이 맞는지 궁금합니다! (아니면 public을 안하더라도 테스트가 가능한 방법이 있을지도 궁금합니다!!)
(실제 코드에서 LottoNumbers 모델의 matches 메소드에 해당합니다)

ilmare-cbk committed:
test(StringAddCalculatorTest.java): 테스트 케이스 추가 (리뷰 반영 사항)

커스텀 구분자로 문자열을 구분하는 테스트에 테스트 케이스 추가

ilmare-cbk committed:
docs(README.md): step3 요구사항 정리

- 로또 생성
- 당첨 통계
- 입력 로직 구현
- 출력 로직 구현
- 컨트롤러 구현

ilmare-cbk committed:
feat(lotto package): 로또번호 생성 로직 구현

- LottoNumber 모델 구현(int형 필드를 wrapping)
- 로또번호를 생성하는 LottoNumberGenerator 모델 구현
- LottoNumberGenerator에서 로또번호 생성 역할을 LottoNumberGeneratorStrategy에
위임
- LottoNumberGeneratorStrategy 구현체에서 로또번호 생성 로직 구현
- 로또번호 생성 테스트 작성(6개의 로또번호, 중복없음, 1 ~ 45 사이의 숫자)

ilmare-cbk committed:
docs(README.md): step3 요구사항 수정

로또번호 오름차순은 도메인 요구사항이기 보다는 출력(view)에 대한 요구사항이라고 판단하여 변경

ilmare-cbk committed:
feat(LottoGenerator.java): 구입할 금액에 맞게 로또를 생성하는 로직 구현

- LottoGenerator에 구입할 금액이 입력되면 로또 1장 가격에 맞게 로또를 생성
- 돈과 관련된 context를 처리하는 Money 모델 구현
- Money를 나눈 값을 확인하는 테스트 작성
- 구입할 금액에 따라 생성된 로또 갯수를 확인하는 테스트 작성

ilmare-cbk committed:
test(LottoGeneratorTest.java): 로또 한 장당 가격이 다를 경우 구입할 수 있는 로또 갯수 확인 테스트 추가

ilmare-cbk committed:
test(LottoNumberGeneratorTest.java): LottoNumber 생성 시 유효성 검사 테스트 추가

- LottoNumber 생성 시 입력된 수가 1 ~ 45가 아니면 IllegalArgumentException을 던진다.
- 공통으로 사용하는 상수를 LottoConstant로 이동

ilmare-cbk committed:
docs(README.md): step3 요구사항 수정

구입한 로또번호와 당첨번호를 비교하는 요구사항 분리
- 로또번호와 당첨번호 1개를 비교하는 기능
- 로또번호와 당첨번호 전체를 비교하는 기능

ilmare-cbk committed:
docs(README.md): step3 요구사항 수정

당첨 여부 판단 요구사항 추가

ilmare-cbk committed:
test(WinTest.java): 당첨번호 비교 및 당첨여부 판단 테스트 작성

- 로또번호(로또 1장)와 당첨번호 1개 비교 테스트
- 로또번호(로또 1장)와 입력된 당첨번호를 전체 비교 테스트
- 로또가 당첨되었는지 확인
- Lotto 모델에 당첨여부를 확인할 수 있는 WinPolicy 주입
- WinPolicy의 구현체(DefaultWinPolicy)를 통해 당첨여부 판단

ilmare-cbk committed:
refactor(lotto package): 정적 팩토리 메서드 이름 변경

ilmare-cbk committed:
refactor(Money.java): Money의 인스턴스 변수 타입을 double로 변경

ilmare-cbk committed:
feat(statistic package): 당첨 통계 기능 구현

- LottoStatistic 모델을 통해 로또 당첨에 대한 통계결과를 LottoSatisticResults로 반환한다.
- 로또 목록과 당첨번호가 주어지면 3~6개의 숫자가 일치한 로또의 갯수를 확인하는 테스트 작성
- 로또 목록과 당첨번호가 주어지면 수익률을 확인하는 테스트 작성

ilmare-cbk committed:
feat(lotto package): 로또 UI(입력,출력) 기능 구현

- 사용자 입력에 대한 유효성 체크 (0이상의 값, 1~45 내의 중복되지 않은 6자리 수)
- 구입한 로또 갯수 및 로또 목록 출력 (오름차순)
- 당첨 통계 출력 (당첨 갯수, 수익률)

ilmare-cbk committed:
refactor(lotto package): 패키지 및 변수명 변경, 불필요한 로직 제거

- 도메인 모델을 model 패키지 하위로 이동
- purchasePrice -> purchaseAmount로 통일
- Lotto 모델의 isWin 메소드 제거 및 WinPolicy 제거 (사용하지 않음)
- Money의 divide 예외 제거 (정상적인 프로그램 동작을 위해 0으로 나눌 때 0으로 리턴하도록 수정)

limwoobin commented on src/main/java/lotto/constant/LottoConstant.java:
```diff
@@ -0,0 +1,11 @@
+    public static final String COMMA_DELIMITER_REGEX = ",";
+    public static final int LOTTO_NUMBER_COUNT = 6;
+
+    private LottoConstant() {
+    }
```
private 으로 객체생성을 막아주셧군요!

ilmare-cbk replied:
유틸성 클래스의 인스턴스화를 막기위해 private 생성자를 추가했습니다!

limwoobin commented on src/main/java/lotto/model/money/Money.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.Objects;
+
+public class Money {
+    private final double value;
```
Money에서 구매금액과 로또의 기본가격인 천원에 대해 같이 처리하고 있군요
다만 Lotto 기본 금액인 천원에 대해서는 Money객체에서 처리할 필요가 있는지는 의문이 드네요 ㅎㅎ
구매금액은 몇장의 로또를 구매하는지, 구매금액에 대해 validation 등의 로직이 필요할텐데
로또 기본금액은 큰 기능이 없어서 이런 경우에는 앞서 작성하신 LottoConstant 에 선언하는게 더 낫지 않을까 생각됩니다!

ilmare-cbk replied:
Money 모델 설계할 때 다음과 같이 생각했었어용!
1. 구매금액이나 기본가격 천원이 모두 같은 context라고 판단해서 Money라는 모델로 표현
2. 기본가격 천원이 바뀔 수 있는 영역이다 라고 판단돼서 상수로 만들지 않았습니다!

그런데 말씀주신 것처럼 구매금액과 기본가격 천원이 분리가 된다면 Money는 구매금액과 관련된 비즈니스 로직만 가져 테스트하기에도 수월할 것 같아요! 👍

limwoobin commented on src/main/java/lotto/model/lotto/LottoNumberGeneratorStrategy.java:
```diff
@@ -0,0 +1,7 @@
+
+import java.util.List;
+
+public interface LottoNumberGeneratorStrategy {
```
로또번호 생성전략에 대해 전략패턴을 이용해 다향성을 고려하셨군요!
의도는 충분히 좋은 의도라고 생각됩니다!
다만, 도메인마다 다르겠지만 로또번호 추출에 대해서는 오버엔지니어링이 될수도 있겠다 라는 생각도 있습니다.
로또번호추출의 방법에 대해 요구사항에 명확히 명시되있어 이런경우엔 추후 새로운 요구사항이 나왔을때 구조를 개선하는것도 좋은방법입니다!
초반부터 과하게 여러 사항을 고려하게 되면 정작 핵심 비즈니스 코드를 해칠수도 있어서요 ㅎㅎ

+ 사실 이런 케이스처럼 어느정도까지 분리하고 설계해야할지가 가장 어렵긴 합니다 ㅎㅎ
그래도 의도는 굉장히 좋네요!

ilmare-cbk replied:
LottoNumberGeneratorStragtegy 을 사용한 이유는 1~45사이의 번호를 추출한다 라는 결과는 같겠지만 추출 성능이 달라질수도 있겠다?라고 판단했어요! 
예를 들어 A Strategy에서 사용하는 로직은 추출하는데 10초가 걸리고, B Strategy는 5초가 걸리는 경우 B Strategy로 교체할 수도 있겠다라고 생각을 한거였어용 ㅎㅎㅎ 
저도 매번 어디까지 고려해서 설계해야되나 항상 고민이었는데 우빈님께서 말씀주신 것처럼 오버엔지니어링은 아닐까 한번더 생각하는 것도 필요할 것 같습니다! 😁

limwoobin replied:
넵 디자인패턴을 이용해 객체를 교체한다는 발상은 굉장히 좋은것 같습니다!
이제 중요한게 상황에 맞게 적용하는건데 정답이 없는문제라 이게 젤 어렵긴합니다 ㅎㅎ

ilmare-cbk replied:
맞아요 어려운 부분인 것 같습니다!
그래도 변화가 있을 수 있는 부분을 인지하고 어떤식으로 처리하면 좋을지 고민해볼 수 있는 좋은 경험이었습니당! ㅎㅎ

limwoobin commented on src/main/java/lotto/model/lotto/DefaultNumberGeneratorStrategy.java:
```diff
@@ -0,0 +1,38 @@
+        this.seedNumbers = createSeedNumbers();
+    }
+
+    private List<Integer> createSeedNumbers() {
+        List<Integer> numbers = new ArrayList<>();
+
+        for (int i = LottoConstant.MIN_LOTTO_NUMBER; i <= LottoConstant.MAX_LOTTO_NUMBER; i++) {
+            numbers.add(i);
+        }
+
+        return numbers;
+    }
```
```
List<Integer> lottoNumbers = IntStream.rangeClosed(1, 45).boxed()
        .collect(Collectors.toList());
```
이런방식으로 static에 선언해 1, 45 의 숫자를 캐싱해서 성능을 챙겨가는건 어떨까요??

ilmare-cbk replied:
오 저런 방법도 있었군용!
말씀해주신 것 처럼 매번 저렇게 생성하지 않고 static으로 선언해 캐싱의 장점을 가져갈 수 있을 것 같습니다!!

limwoobin commented on src/main/java/lotto/model/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,51 @@
+import lotto.constant.LottoConstant;
+import lotto.message.ErrorMessages;
+
+public class LottoNumber implements Comparable<LottoNumber> {
```
로또번호까지 일급객체로 감싸서 사용하셨군요!
객체를 역할에 맞게 나누는모습이 굉장히 인상적입니다 ㅎㅎ

ilmare-cbk replied:
일급컬렉션과 일급객체를 사용하면 관련된 로직들을 한 곳에서 처리할 수 있어, 정말 매력적이라고 생각합니다 ㅎㅎㅎ

limwoobin commented on src/main/java/lotto/model/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,51 @@
+        return Objects.hash(number);
+    }
+
+    @Override
+    public int compareTo(LottoNumber o) {
+        return Integer.compare(this.number, o.number);
+    }
```
comparable 을 상속받아 compareTo 메서드를 오버라이딩하신 이유가 있을까요??
로또번호의 경우 수가 같은지 비교하지 크고작은것에 대해서는 사용할 일이 없을것같아서요!

ilmare-cbk replied:
요구사항 중에 로또번호 정렬이 있어 compareTo 메서드를 재정의 했습니다!
요구사항 힌트에 Collections.sort()를  확인하지 못했네요 😅
재정의한 부분 제거하고 Collections.sort로 수정해보겠습니다!

ilmare-cbk replied:
Collections.sort()에 넘기는 파라미터가 Comparable 를 구현한 element를 갖는 collection이어야 해서 compareTo 메서드는 그대로 유지하게 되었습니당!

limwoobin replied:
넵 이부분은 제가 잘못말씀드렸네요 ㅠ

limwoobin commented on src/main/java/lotto/model/lotto/LottoNumbers.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class LottoNumbers {
```
Lotto 객체가 있는데 굳이 LottoNumbers가 있을필요가 있을까요??
해당 비즈니스로직은 Lotto에서 담당해도 충분하지 않을까 생각드네요 ㅎㅎ

ilmare-cbk replied:
맞습니다.. ㅎㅎㅎ 저도 이부분이 계속 맘에 걸렸는데 다시 한번 고민해보고 변경해보도록 하겠습니다!

ilmare-cbk replied:
우빈님 말씀하신대 Lotto가 List<LottoNumber>를 갖도록 수정했습니다!
수정하고 나니 불필요한 클래스들을 제거할 수 있어서 설계적인 측면에서 한결 깔끔해졌습니다 ㅎㅎㅎ
감사합니다 👍

limwoobin commented on src/main/java/lotto/model/lotto/LottoNumbers.java:
```diff
@@ -0,0 +1,51 @@
+        return this.values.contains(lottoNumber);
+    }
+
+    @Override
+    public String toString() {
+        values.sort(LottoNumber::compareTo);
+        return values.toString();
+    }
```
앞서 리뷰남겼던 compareTo가 여기 있었군요 ㅎㅎ
요구사항의 힌트를 보시면 `Collections.sort() 메소드를 활용해 정렬 가능하다.` 라고 있습니다
Collections의 기능을 사용하시는건 어떨까요?

ilmare-cbk replied:
ㅎㅎㅎ 요구사항 힌트를 조금 더 잘 살펴봐야겠습니다😅
해당 부분은 Collections.sort를 사용해서 수정해보겠습니다!!

limwoobin commented on src/main/java/lotto/model/lotto/LottoGenerator.java:
```diff
@@ -0,0 +1,35 @@
+        return new LottoGenerator(lottoNumberGenerator, priceOfOneLotto);
+    }
+
+    public Lottos generate(Money purchaseAmount) {
+        double count = purchaseAmount.divide(priceOfOneLotto);
+        return generate((int) count);
+    }
```
이 경우 Money로 수량을 나누기보단 LottoConstant에 로또 기본금액을 선언해 나누는것이 더 명확하겠네요!

ilmare-cbk replied:
네!  기본금액을 상수로 변경하고 Money의 역할을 명확하게 가져가는게 좋을 것 같습니다!

limwoobin commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -0,0 +1,40 @@
+import lotto.util.LottoInputValidator;
+import lotto.util.StringToIntegerConvertor;
+
+public class InputView {
```
view 영역으로 나누신거 좋네요!

ilmare-cbk replied:
감사합니다 😄

limwoobin commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -0,0 +1,40 @@
+    private InputView() {
+    }
+
+    public static int inputPurchaseAmount() {
+        System.out.println(LottoMessage.INPUT_PURCHASE_AMOUNT);
+        String purchaseAmount = SCANNER.nextLine();
+
+        if (LottoInputValidator.validateNonNegativeNumber(purchaseAmount)) {
+            return Integer.parseInt(purchaseAmount);
+        }
+
+        System.out.printf((LottoMessage.INVALID_PURCHASE_AMOUNT) + "%n", purchaseAmount);
+        return inputPurchaseAmount();
+    }
+
+    public static List<Integer> inputWinningNumbers() {
+        System.out.println(LottoMessage.INPUT_WINNING_NUMBERS);
+        String input = SCANNER.nextLine();
+
+        if (LottoInputValidator.validateWinningNumbers(input)) {
+            return StringToIntegerConvertor.convertNumbers(input.split(LottoConstant.COMMA_DELIMITER_REGEX));
+        }
+
+        System.out.printf((LottoMessage.INVALID_WINNING_NUMBERS) + "%n", input);
+        return inputWinningNumbers();
+    }
```
validate를 선언하는 위치에 대해서는 한번 같이 고민해보면 좋을것같습니다!
input쪽에 선언하는것도 있지만 객체내에서도 선언하는것을 한번 고려해보시져 ㅎㅎ 

저같은 경우엔 코드를 작성할때 넘어오는값에 대해 클라이언트를 믿지 않는 편입니다(언제든 잘못된 값이 넘어올 수 있다고 가정합니다 ㅎㅎ)
같이 협업하는 개발자가 있는 경우라면 LottoNumber에 잘못된 값을 선언해버려 로직이 잘못 동작하게 되는 경우를 생각해볼 수 있습니다.
이런 여러 경우를 고려하여 input도 좋지만 저는 객체내에도 같이 validate 로직을 작성하는편입니다!

ilmare-cbk replied:
항상 정상적인? 객체들만 생성될 수 있도록 객체 내에서의 validate가 꼭 필요할 것 같습니다!

limwoobin commented on src/main/java/lotto/model/statistic/LottoStatistic.java:
```diff
@@ -0,0 +1,45 @@
+import lotto.model.money.Money;
+import lotto.model.win.WinRanking;
+
+public class LottoStatistic {
```
statistic package의 역할이 애매한거같기도 하네요 ㅎㅎ
수익률 계산을 위해서라면 Profit관련 객체를 하나 생성하시는것도 좋은 방법이 될 수 있습니다!

ilmare-cbk replied:
수익률과 관련된 로직은 Profit이라는 모델로 처리할 수 있을 것 같습니다!
statistic 패키지 역할에 대해서는 다시 한번 고민해보도록 하겠습니당!!

ilmare-cbk replied:
리팩토링을 통해 statistic 패키지와 하위 클래스들은 제거 될 수 있었습니다.
굳이 statistic 관련 클래스들이 없더라도 Lottos 도메인을 통해 LottoStatistic이 할 수 있는 역할들을 처리할 수 있었고,
수익률 계산도 profit  도메인 모델을 두어 수익률과 관련된 로직만 담도록 수정했습니다!
확실히 역할 맞게 도메인을 분리하다 보니 코드는 명확해지고, 불필요한 요소들은 제거돼서 좋았습니다!!

limwoobin commented on src/main/java/lotto/Application.java:
```diff
@@ -0,0 +1,16 @@
+
+public class Application {
+    public static void main(String[] args) {
+        LottoController lottoController = new LottoController(
+                LottoGenerator.of(LottoNumberGenerator.from(new DefaultNumberGeneratorStrategy()), Money.from(1000))
+        );
```
객체생성시 정적팩토리, 생성자를 여러개 조합해서 한번에 생성하셧군요
이런 방식은 뎁스가 깊어져 가독성을 떨어트릴 수 있습니다.

LoginController 생성자 내에서 직접 처리해보는건 어떠실까요??

ilmare-cbk replied:
아 미처 거기까지 생각하지 못했네요 ㅎㅎ
말씀주신 것 처럼 생성자 내에서 처리하면 가독성도 높이고 코드도 깔끔해질 것 같네요!👍

limwoobin commented on src/test/java/lotto/TestLottoNumberGeneratorStrategy.java:
```diff
@@ -0,0 +1,30 @@
+import lotto.model.lotto.LottoNumber;
+import lotto.model.lotto.LottoNumberGeneratorStrategy;
+
+public class TestLottoNumberGeneratorStrategy implements LottoNumberGeneratorStrategy {
```
`TestLottoNumberGeneratorStrategy` 는 테스트에서만 사용되는 객체인것같은데 `LottoNumberGeneratorStrategy` 인터페이스를 상속받고 있군요 ㅎㅎ
이 방식은 testcode와 production가 디펜던시가 강하게 엮여있다고 볼 수 있습니다.
테스트코드는 말그대로 프로덕션코드를 검증하는 코드라서 둘 사이의 디펜던시는 좋지 않다고 생각합니다.
테스트코드에서 로또번호생성을 위해서라면 직접 LottoNumber를 생성하시는것이 어떨까 의견드려요 ㅎㅎ

ilmare-cbk replied:
testcode와 production 사이의 디펜던시가 강해진다 까지 생각을 못했습니다!
테스트코드에서 LottoNumber를 직접 생성하는 것으로 수정하겠습니다 ㅎㅎ
이번 리뷰로 정말 많이 알아가는 것 같습니다
감사합니다!!

limwoobin reviewed (changes requested):
안녕하세요 보경님! 3단계 미션 잘 구현해주셨네요 ㅎㅎ
로또, 로또번호들에 대해 일급객체로 잘게 나눠주신 부분이 인상적이었습니다.
객체를 나누고 검증하는것이 굉장히 애마하고도 어려운데 잘 해내주셨습니다!
다만 고민해보면 좋을 부분들이 있어 Request Change 드렸습니다!

추가로 몇가지 의견을 코멘트로 남겨두었으니 확인부탁드립니다.
model 이라는 패키지명보다는 내부 객체들이 도메인의 성격이 강해 domain 이라는 패키지명은 어떨지 의견드립니다.
그리고 model 패키지아래에 해당 객체들에 대해서는 테스트코드 작성이 안되있어 아쉽네요 ㅎㅎ
기본적으로 객체단위로 테스트코드를 작성해보시는것을 추천드립니다!

질문에 답변드리겠습니다 ㅎㅎ
테스트를 위해 public으로 공개하는것은 저는 지양합니다.
그렇다고 해당 메서드를 위해 리플렉션을 사용하는것 또한 지양하구요
해당 기능을 테스트해야 한다면 public으로 공개된 메서드를 테스트할때 private 에 있는 기능들 까지 고려해서 테스트 케이스를 작성하신다면 예외case, happy case 모두  검증할 수 있을겁니다!ㅎㅎ

(궁금하거나 의견나누고싶은 부분은 코멘트나 dm으로 편히 연락주세요!)

ilmare-cbk committed:
refactor(model package): model 패키지를 domain 패키지로 이름 변경 및 generator 패키지 추가

- model패키지 하위에 있는 클래스들이 도메인의 성격을 나타내고 있어서 domain이라고 명확하게 패키지명을 변경
- generator 패키지를 추가해 generator 관련 클래스들을 해당 패키지로 이동

ilmare-cbk committed:
refactor(LottoNumbers.java): 테스트를 위해 public으로 공개된 메소드를 private으로 변경

private으로 변경하고 private 메소드를 사용하는 public 메소드를 통해 기능 테스트가 되도록 수정

ilmare-cbk committed:
refactor(Money.java): Money 주요 로직 추가

- Money가 구매금액에 대한 개념만 가지도록 변경 (기본금액 1000원 상수화)
- Money 생성 시 구매금액에 대한 유효성 검사 및 테스트 추가
- 구매할 수 있는 수량을 반환하는 로직 추가 및 테스트 추가

ilmare-cbk committed:
refactor(DefaultNumberGeneratorStrategy.java): 한 번 만든 1 ~ 45 숫자 목록을 static하게 설정하여 캐싱되도록 수정

ilmare-cbk committed:
refactor(LottoNumbers.java): 로또번호를 Collections.sort() 를 통해 정렬되도록 수정

ilmare-cbk committed:
refactor(LottoNumbers.java): Lotto를 List<LottoNumber>를 관리하는 일급컬렉션으로 변경

- 기존 LottoNumbers 일급컬렉션을 제거
- LottoNumbers가 가지고 있었던 비즈니스 로직을 Lotto로 이동

ilmare-cbk committed:
refactor(domain.lotto package): Lotto, LottoNumber 생성 시 유효성 검사

- LottoNumber 생성 시 1 ~ 45 이내의 숫자인지 유효성 검사 및 테스트 작성
- Lotto 생성 시 6개의 중복되지 않은 숫자인지 유효성 검사 및 테스트 작성

ilmare-cbk committed:
refactor(LottoInputValidator.java): 사용자 입력값 유효성 검사 로직 수정 및 테스트 작성

- 구매금액이 0이상의 숫자인지 확인
- 구매금액이 로또 한 장 가격이상인지 확인
- 당첨번호가 1 ~ 45 이내의 중복되지 않는 6개의 숫자인지 확인

ilmare-cbk committed:
refactor(lotto package): purchase amount 네이밍을 purchase price 로 변경

ilmare-cbk committed:
refactor(LottoController.java): LottoController 생성자 시그니처 변경

- LottoNumberGeneratorStrategy만 파라미터로 넘겨 LottoController 생성 시 가독성을 높임

ilmare-cbk committed:
refactor(TestLottoNumberGeneratorStrategy.java): LottoNumberGeneratorStrategy implements 제거

- LottoNumberGeneratorStrategy implements 제거해서 production 코드와의 디펜던시 제거
- TestLottoNumberGeneratorStrategy를 TestLottoNumberGenerator로 이름 변경

ilmare-cbk committed:
fix(domain.lotto, util package): 6개의 중복되지 않는 숫자 유효성 검사 로직 수정

- 6개의 숫자를 판단하는 로직과 중복이 있는지 판단하는 로직을 분리

ilmare-cbk committed:
refactor(statistic package): statistic 패키지 및 하위 도메인 모델 제거

- LottoStatistic에서 당첨된 로또 갯수를 판단하는 로직을 Lottos로 변경
- LottoStatistic에서 수익률 계산하는 로직을 Profit 모델로 변경
- Profit의 수익률 및 수익률이 1보다 작은지 판단하는 테스트 작성

ilmare-cbk committed:
refactor(Money.java): 불필요한 메소드 및 import 제거

ilmare-cbk committed:
fix(LottoInputValidator.java): 공백 포함하여 당첨금액 입력하더라도 정상처리되도록 수정 (trim)

ilmare-cbk committed:
test(StringToIntegerConvertorTest.java): 문자열 타입의 숫자 입력시 정상적으로 숫자타입으로 변환되는지 테스트

ilmare-cbk committed:
docs(README.md): step 요구사항 수정

리팩토링된 사항들 반영

ilmare-cbk commented:
> 안녕하세요 보경님! 3단계 미션 잘 구현해주셨네요 ㅎㅎ 로또, 로또번호들에 대해 일급객체로 잘게 나눠주신 부분이 인상적이었습니다. 객체를 나누고 검증하는것이 굉장히 애마하고도 어려운데 잘 해내주셨습니다! 다만 고민해보면 좋을 부분들이 있어 Request Change 드렸습니다!
> 
> 추가로 몇가지 의견을 코멘트로 남겨두었으니 확인부탁드립니다. model 이라는 패키지명보다는 내부 객체들이 도메인의 성격이 강해 domain 이라는 패키지명은 어떨지 의견드립니다. 그리고 model 패키지아래에 해당 객체들에 대해서는 테스트코드 작성이 안되있어 아쉽네요 ㅎㅎ 기본적으로 객체단위로 테스트코드를 작성해보시는것을 추천드립니다!
> 
> 질문에 답변드리겠습니다 ㅎㅎ 테스트를 위해 public으로 공개하는것은 저는 지양합니다. 그렇다고 해당 메서드를 위해 리플렉션을 사용하는것 또한 지양하구요 해당 기능을 테스트해야 한다면 public으로 공개된 메서드를 테스트할때 private 에 있는 기능들 까지 고려해서 테스트 케이스를 작성하신다면 예외case, happy case 모두 검증할 수 있을겁니다!ㅎㅎ
> 
> (궁금하거나 의견나누고싶은 부분은 코멘트나 dm으로 편히 연락주세요!)

---
- 도메인 성격에 맞게 model 패키지를 domain 패키지로 이름 변경했습니다!
- 도메인 모델의 누락된 테스트들도 추가로 작성했습니다!
- public으로 공개해서 테스트하는 메소드 없이 public 메소드를 통해 private 메소드도 같이 테스트 받도록 수정했습니다!

감사합니다😃

limwoobin commented on src/main/java/lotto/generator/DefaultNumberGeneratorStrategy.java:
```diff
@@ -0,0 +1,29 @@
+
+public class DefaultNumberGeneratorStrategy implements LottoNumberGeneratorStrategy {
+
+    private static final List<Integer> SEED_NUMBERS =
```
캐싱처리 좋습니다!

limwoobin commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,34 @@
+
+        ResultView.printPurchaseResult(lottos);
+
+        List<LottoNumber> winningNumbers =
+                InputView.inputWinningNumbers()
+                        .stream()
+                        .map(LottoNumber::from)
+                        .collect(Collectors.toList());
```
당첨번호에 대해서는 컬렉션을 사용하셨네요
Lotto 객체와 동일한 필드를 가질거같은데 Lotto를 사용하지 않은 이유가 있을까요??
혹은 역할분리를 위해 당첨번호 객체를 따로 만드는것도 고려해보면 좋을것같습니다!

ilmare-cbk replied:
Lotto는 구입한 로또번호만을 관리하는 일급컬렉션이라고 생각했습니다
그래서 당첨번호에 대해서는 Lotto를 사용하지 않았는데, 말씀주신 것 처럼 당첨번호만을 위한 객체가 필요한지 확인해보고,
필요하다면  해당 객체를 통해 도메인 로직을 처리할 수 있도록 수정해보겠습니다!!
리뷰 해주셔서 감사합니다 👍

ilmare-cbk replied:
@limwoobin 앗 혹시 당첨번호 객체를 따로 만들라고 말씀하신게 일급컬렉션을 말씀하시는게 맞을까요? ㅎㅎ

limwoobin replied:
당첨번호에 대한 객체가 따로있을때 이점이 있다고 생각되시면
당첨번호 객체를 추가하는것을 고민해보시라는 뜻이었습니다 ㅎㅎ

limwoobin reviewed (approved):
보경님 수정해주신 3단계 리뷰 잘봤습니다!
코멘트 드린내용 잘 수정해주셨네요 ㅎㅎ

간단히 코멘트 몇가지 남겨드렸습니다. 다음 미션하면서 확인부탁드려요
이번 미션은 여기서 merge하겠습니다 고생많으셨어요
다음 단계도 화이팅입니다
